pkg/reconciler/source/crd/crd.go

```diff
@@ -0,0 +1,308 @@
+	// The reconciliation process is as follows:
+	// 	1. Resolve GVR and GVK from a particular Source CRD (i.e., those labeled with duck.knative.dev/source = "true")
+	//  2. Dynamically create a controller for it, if not present already. Such controller is in charge of reconciling
+	//     duckv1.Source resources with that particular GVR..
+
+	gvr, gvk, err := r.resolveGroupVersions(ctx, crd)
+	if err != nil {
+		logging.FromContext(ctx).Error("Error while resolving GVR and GVK", zap.String("CRD", crd.Name), zap.Error(err))
+		return err
+	}
+
+	err = r.reconcileController(ctx, crd, gvr, gvk)
+	if err != nil {
+		logging.FromContext(ctx).Error("Error while reconciling controller", zap.String("GVR", gvr.String()), zap.String("GVK", gvk.String()), zap.Error(err))
+		return err
+	}
+
+	return nil
+}
+
+func (r *Reconciler) finalize(ctx context.Context, crd *v1beta1.CustomResourceDefinition) error {
+	// The Source CRD is being deleted, we need to delete the dynamically created controller...
+	gvr, gvk, err := r.resolveGroupVersions(ctx, crd)
+	if err != nil {
+		logging.FromContext(ctx).Error("Error while resolving GVR", zap.String("CRD", crd.Name), zap.Error(err))
+		return err
+	}
+	r.lock.RLock()
+	rc, found := r.controllers[*gvr]
+	r.lock.RUnlock()
+	if found {
+		r.lock.Lock()
+		// Now that we grabbed the write lock, check that nobody deleted it already.
+		rc, found = r.controllers[*gvr]
+		if found {
+			logging.FromContext(ctx).Info("Stopping Source Duck Controller", zap.String("GVR", gvr.String()), zap.String("GVK", gvk.String()))
+			rc.cancel()
+			delete(r.controllers, *gvr)
+		}
+		r.lock.Unlock()
+	}
+	return nil
+}
+
+func (r *Reconciler) resolveGroupVersions(ctx context.Context, crd *v1beta1.CustomResourceDefinition) (*schema.GroupVersionResource, *schema.GroupVersionKind, error) {
+	var gvr *schema.GroupVersionResource
+	var gvk *schema.GroupVersionKind
+	for _, v := range crd.Spec.Versions {
+		if !v.Served {
+			continue
+		}
+		gvr = &schema.GroupVersionResource{
+			Group:    crd.Spec.Group,
+			Version:  v.Name,
+			Resource: crd.Spec.Names.Plural,
+		}
+
+		gvk = &schema.GroupVersionKind{
+			Group:   crd.Spec.Group,
+			Version: v.Name,
+			Kind:    crd.Spec.Names.Kind,
+		}
+
+	}
+	if gvr == nil || gvk == nil {
+		return nil, nil, fmt.Errorf("unable to find GVR or GVK for %s", crd.Name)
+	}
+	return gvr, gvk, nil
+}
+
+func (r *Reconciler) reconcileController(ctx context.Context, crd *v1beta1.CustomResourceDefinition, gvr *schema.GroupVersionResource, gvk *schema.GroupVersionKind) error {
+	r.lock.RLock()
+	rc, found := r.controllers[*gvr]
+	r.lock.RUnlock()
+	if found {
+		return nil
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	// Now that we grabbed the write lock, check that nobody has created the controller.
+	rc, found = r.controllers[*gvr]
+	if found {
+		return nil
+	}
+
+	// Source Duck controller constructor
+	sdc := duck.NewController(crd.Name, *gvr, *gvk)
```

Comment:
ooh you are making a controller. cool. nevermind. this is like autotrigger.

Reply:
yeah, and that is why I kind of have the finalizer to remove this controllers if something gets removed.. but you prefer not to patch CRDs right?

Reply:
worst case the controller throws errors because a missing CRD to GVK lookup and we can handle that then, no need to block deleting the CRD.

Reply:
OK agree